sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -95,9 +95,14 @@ private[sql] object Dataset {
   def ofRows(sparkSession: SparkSession, logicalPlan: LogicalPlan): DataFrame =
     sparkSession.withActive {
       val qe = sparkSession.sessionState.executePlan(logicalPlan)
-      qe.assertAnalyzed()
-      new Dataset[Row](qe, RowEncoder.encoderFor(qe.analyzed.schema))
-  }
+      val encoder = if (qe.isLazyAnalysis) {
+        RowEncoder.encoderFor(new StructType())
```

Comment:
@ueshin this breaks collect (and other operations) on these dataframes. The alternative would be to defer the construction of the encoder until it is needed.

Reply:
@hvanhovell sure, submitted #48829.